### ISSUE TITLE
[DFA-arch T0] MLP classifier at small output dim — the positive-baseline ANCHOR (harness sanity: does toy reproduce a DFA win at all?)

### ISSUE THREAD

**Issue by OriPekelman** [OriPekelman/toy#152]
**Tier 0 (do first).** The DFA-vs-BP program has thoroughly shown DFA FAILS for transformer LMs (huge output dim, vocab 50257) but has NO positive baseline anywhere. Our mechanistic finding (F13/F13b) + theory (Refinetti et al., "Align, then memorise", ICML 2021, arXiv:2011.12428) say DFA works when the OUTPUT DIMENSION is small. The cheapest place to prove that — and to sanity-check that our harness can reproduce a *known* DFA success — is a plain MLP classifier at low #classes (DFA's most reliable near-BP regime; MNIST-parity in the literature, Nokland 2016 arXiv:1609.01596).

## Ask
A minimal **MLP classifier** path: N-layer MLP + a small-#classes head + cross-entropy on a small dataset (synthetic gaussians / tabular / a small image set — whatever is cheapest to wire). Must support the existing per-layer `--policy chain|dfa` + `--policy-scope`, so we can run all-BP vs all-DFA. Emit the existing align + loss/accuracy events.

## Success target (the anchor)
all-DFA within a small gap of all-BP on val accuracy at low #classes (~10). If it's near-BP -> we finally have a positive, our output-dim lens is validated as a POSITIVE (not just a failure explanation), and every later cross-arch bet has a working reference. Sweep #classes {2, 10, 100, 1000} to trace the output-dim degradation directly (should worsen with #classes — our lens, measured).

## Consuming Tao experiment: F16 (anchor).
Refs: Nokland 2016; Refinetti 2021; Bartunov et al. NeurIPS 2018 (where plain FA stops scaling).

**Comment by OriPekelman** [comment]
Shipped in 5c77415 — and **the anchor holds: this is the program's first DFA positive.**

`toy train mlp` (runner `libexec/toy-train-mlp`, engine `lib/toy/llm/engine/mlp_engine.rb`, gate `prep/mlp_gate.rb`). Per-hidden-layer `--policy chain|dfa|frozen`; the head always trains by BP (at the output layer DFA and BP coincide). CPU-only, own compilation unit, `MLP_*` env namespace, canonical `g_l = a_inᵀ(B_l e)ᵀ` with `B_l` shaped `[n_classes, d_out]` off the shared `Toy::Train::DfaB`.

## The measurement

3×64 hidden / 32 features / batch 64 / 1000 steps / lr 0.003, matched init + seed, 2048 held-out samples (seed 0 — the numbers the gate asserts):

| classes | all-BP | all-DFA | frozen | recovery |
|---------|--------|---------|--------|----------|
| 2       | 0.940  | **0.942** | 0.772 | 1.01 |
| 10      | 0.773  | 0.742   | 0.516  | 0.88 |
| 100     | 0.559  | 0.412   | 0.273  | 0.49 |
| 1000    | 0.411  | 0.265   | 0.162  | 0.42 |

*recovery* = (DFA − frozen)/(BP − frozen) — the share of BP's over-the-frozen-control gain DFA recovers. Raw accuracy is not comparable across output dims (chance is 1/C); this is.

1. **A POSITIVE.** At 2 classes DFA matches BP; at 10 it is 3 points behind BP and 23 ahead of the frozen control, clearing the mandatory bar on both halves. So the F4–F14 negatives are findings about transformer LMs at vocab 50257, **not harness artifacts** — which was this ticket's real job.
2. **The output-dim lens, MEASURED.** Recovery falls monotonically 1.01 → 0.88 → 0.49 → 0.42 across {2, 10, 100, 1000}, on every seed tried (0/1/2), on the same feedback machinery the franken lanes use. The lens was a failure *explanation*; it is now a *tested prediction*.
3. **The mechanism is visible.** With `--align-events`, cos(g_DFA, g_BP) climbs from ~0: w1 −0.002 → 0.67, w2 −0.089 → 0.60, w3 −0.04 → 0.31 (seed 0, 1000 steps). "Align, then memorise", reproduced — and note w3, nearest the head, aligns *least*: the same depth signature the F-series saw. The gate asserts the climb, because a broken DFA wiring can still move a loss curve but cannot make the shadow gradient rotate towards a random matrix.

## Three deviations from the ticket text — flagged, not silent

- **The task is a random-teacher network, not gaussian blobs** (`--task blobs` still ships, and its degeneracy is *measured*: BP 1.000 / DFA 0.9995 / **frozen 0.992**). Blobs are linearly separable, so the frozen control ties and the mandatory frozen-beat could never be met by *anything* — the anchor would have been vacuous by construction, i.e. exactly the failure mode tao#19's bar was written to catch. Gaussian inputs + a fixed random ReLU teacher is Refinetti et al.'s own setup, keeps every property asked for (synthetic, seeded, no plumbing), and makes the hidden layers load-bearing.
- **The DFA rule keeps the ⊙f′(a) factor** the franken lane's per-matmul surrogate drops (`ggml_silu_back` gives the exact derivative). Had the anchor failed with the surrogate, we could not have told "DFA does not work at small output dim" from "our surrogate is not DFA".
- **Train/val disjointness is by construction**, not by seed luck: the val set is materialised from the head of one sample stream, training continues from what follows. Two differently-seeded streams are two offsets into the *same* LCG cycle and can silently overlap.

## Settled decisions honoured

`--policy-scope` is **rejected** on this recipe (tao#18 item 1; pinned in `train_cli_matrix_gate`), CPU-only with no CUDA twin (tao#18 item 2), align events keep `li` / lane-local `wi` / the new **`wname`** string (tao#19 item 3), and the success bar is baked into the gate (tao#19 item 4). run_start carries a **`dfa`** object, *not* `franken` — a consumer keying on `franken` would read an MLP run as a transformer one.

## For F16

```
toy train mlp --steps 1000 --seed 0 --classes 10 --val-batches 32 --policy chain,chain,chain
toy train mlp --steps 1000 --seed 0 --classes 10 --val-batches 32 --policy dfa,dfa,dfa --align-events
toy train mlp --steps 1000 --seed 0 --classes 10 --val-batches 32 --policy frozen,frozen,frozen
```

Arms differ **only** in `--policy`. ~0.5 s per cell. `val: acc=… loss=… n=…` on stdout; events carry per-step `train_acc`, the `align` events, and one end-of-run `eval`.

**Two things to know before running a sweep.** (1) At `--classes 100+`, pass `--val-batches 32`: at the 512-sample default the 1000-class recovery is noise-dominated and the 100→1000 comparison flipped sign between two honest val sets. (2) Under AdamW the **B scale axis barely moves this lane** (inv_sqrt_fan / glorot / fixed:1.0 within a point) — Adam normalises per-parameter magnitude, so only B's *direction* carries information. Don't spend a sweep on it. DFA does, however, tolerate less LR than BP (val loss blows up at 0.03 while BP is unbothered) — hence this lane's own 0.003 default.

47 gate targets green (55 legs; `gate-mlp` new). Full write-up in `docs/roadmap/dfa-arch-program-2026-08-10.md`.

**Comment by OriPekelman** [comment]
Closing: shipped in `5c77415` — feat(dfa-arch): toy#152 T0 anchor — `toy train mlp`, and DFA's FIRST POSITIVE

Verified by `git log`; the implementing commit names this issue in its subject. Part of the 2026-08-14 issue-state sweep (48 closed, all against a commit in `main`). Reopen if the scope here outran what landed.
